misc/peer-store/src/store.rs

```diff
@@ -0,0 +1,99 @@
+use std::{
+    collections::{HashMap, HashSet},
+    time::SystemTime,
+};
+
+use libp2p_core::{Multiaddr, PeerId};
+
+/// A store that
+/// - keep track of currently connected peers;
+/// - contains all observed addresses of peers;
+pub trait Store {
+    /// Called when a peer connects.
+    fn on_peer_connect(&mut self, peer: &PeerId);
+    /// Called when a peer disconnects.
+    fn on_peer_disconnect(&mut self, peer: &PeerId);
+    /// Update an address record.  
+    /// Return `true` when the address is new.  
+    fn on_address_update(&mut self, peer: &PeerId, address: &Multiaddr) -> bool;

```

Comment:
Nit: `update` gives the impression that a peer just has one address, and that this address gets updated here.
Wdyt of instead of calling it instead `on_new_address` or something like that?

Reply:
Because for now I only see one address pop up at a time. I was planning to use a boxed slice but you know there would be a heap allocation, which isn't necessary for only one element.

Reply:
Also we can't do a batch update so there will be a iterator anyway.

Reply:
Sorry, maybe I wasn't clear. I was just nitpicking on the name of the function, not the `address: &Multiaddr` parameter :)

Reply:
Um, I'm not so sure about the naming, because the address isn't necessarily a new address. If the address is not new, it is updated due to LRU rules, like `touch`, so I can't quite make the decision.